### PR TITLE
Lightning Lite docs

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #9987 
This should be merged after #10175 

This is the V1 for the new Lightning Lite package. Planned to be released as part of 1.5.

### Demo
Lightning Lite Demo

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

### CODE REVIEW

File: docs/source/starter/lightning_lite.rst [hunk 1]
```diff
@@ -0,0 +1,660 @@
...
+   * - Batch-level fault tolerance training
+     - Ability to resume from a failure as if it never happened.
+
+
+If you are facing one of those challenges then you are already meeting the limit of :class:`~pytorch_lightning.lite.LightningLite`.
+We recommend you to convert to :doc:`Lightning <../starter/new-project>`, so you never have to worry about those.
+
+----------
+
+Convert to Lightning
```
[the remark this answers is not in the document]
Comment (tchaton) [reply docs/source/starter/lightning_lite.rst:224]: Hey @SkafteNicki. Good point, I would try to add maybe an extra step to demonstrate the convertion.

File: docs/source/starter/lightning_lite.rst [hunk 2]
```diff
@@ -0,0 +1,721 @@
...
+Convert to LightningLite
+========================
+
+Here are 5 required steps to convert to :class:`~pytorch_lightning.lite.LightningLite`.
+
+1. Subclass :class:`~pytorch_lightning.lite.LightningLite` and override its :meth:`~pytorch_lightning.lite.LightningLite.run` method.
+2. Move the body of your existing `run` function.
+3. Remove all calls to `to`, `cuda` etc. as `LightningLite` will take care of it
+4. Apply :meth:`~pytorch_lightning.lite.LightningLite.setup` over each model and optimizers pair, :meth:`~pytorch_lightning.lite.LightningLite.setup_dataloaders` on all your dataloaders and replace ``loss.backward()`` by ``self.backward(loss)``
+5. Instantiate your :class:`~pytorch_lightning.lite.LightningLite` and call its :meth:`~pytorch_lightning.lite.LightningLite.run` method.
```
Comment (rohitgr7) [inline docs/source/starter/lightning_lite.rst:86]:
```suggestion
1. Subclass :class:`~pytorch_lightning.lite.LightningLite` and override its :meth:`~pytorch_lightning.lite.LightningLite.run` method.
2. Move the body of your existing `run` function.
3. Remove all `.to`, `.cuda` etc calls since `LightningLite` will take care of it.
4. Apply :meth:`~pytorch_lightning.lite.LightningLite.setup` over each model and optimizers pair, :meth:`~pytorch_lightning.lite.LightningLite.setup_dataloaders` on all your dataloaders and replace ``loss.backward()`` by ``self.backward(loss)``.
5. Instantiate your :class:`~pytorch_lightning.lite.LightningLite` and call its :meth:`~pytorch_lightning.lite.LightningLite.run` method.
```

File: docs/source/starter/lightning_lite.rst [hunk 3]
```diff
@@ -0,0 +1,721 @@
...
+    Lite(strategy="ddp", gpus=8, num_nodes=32, accelerator="gpu").run()
+
+
+.. note:: We recommend instantiating the models within the :meth:`~pytorch_lightning.lite.LightningLite.run` method as large models would cause an out-of-memory error otherwise.
+
+.. note::
+
+    If you have hundreds or thousands of line within your :meth:`~pytorch_lightning.lite.LightningLite.run` function
+    and you are feeling weird about it, this is exactly the expected feeling.
+    This is how our :class:`~pytorch_lightning.core.lightning.LightningModule` started 2-3 years ago
```
Comment (rohitgr7) [inline docs/source/starter/lightning_lite.rst:194]: maybe mention the year since `2-3 years` is relative.
```suggestion
    This is how our :class:`~pytorch_lightning.core.lightning.LightningModule` started in 2019
```

File: docs/source/starter/lightning_lite.rst [hunk 4]
```diff
@@ -0,0 +1,721 @@
...
+   * - Cross processes reduction
+     - Wrongly reported metrics or gradients due mis-reduction.
+   * - Large sharded models
+     - Instantiation, materialization and state management of large models.
+   * - Rank 0 only actions
+     - Logging, profiling, etc.
+   * - Checkpointing / Early stopping / Callbacks
+     - Ability to easily customize your training behaviour and make it stateful.
+   * - Batch-level fault tolerance training
+     - Ability to resume from a failure as if it never happened.
```
Comment (rohitgr7) [inline docs/source/starter/lightning_lite.rst:223]:
```suggestion
   * - Processes divergence
     - This happens when processes execute a different section of the code due to different if/else conditions, race condition on existing files, etc., resulting in hanging.
   * - Cross processes reduction
     - Wrongly reported metrics or gradients due to mis-reduction.
   * - Large sharded models
     - Instantiation, materialization and state management of large models.
   * - Rank 0 only actions
     - Logging, profiling, etc.
   * - Checkpointing / Early stopping / Callbacks / Logging
     - Ability to easily customize your training behaviour and make it stateful.
   * - Batch-level fault tolerance training
     - Ability to resume from a failure as if it never happened.
```

File: docs/source/starter/lightning_lite.rst [hunk 5]
```diff
@@ -0,0 +1,721 @@
...
+    class Lite(LightningLite):
+
+        # 1. This would becomes the LightningModule `__init__` function.
+
+        def run(self, args):
+            self.args = args
+
+            model = MyModel(...)
+
+            self.fit(model)  # This would be automated by Lightning Trainer.
```
Comment (rohitgr7) [inline docs/source/starter/lightning_lite.rst:254]:
```suggestion
            self.model = MyModel(...)

            self.fit()  # This would be automated by Lightning Trainer.
```

File: docs/source/starter/lightning_lite.rst [hunk 6]
```diff
@@ -0,0 +1,721 @@
...
+
+        # 2. This can be fully removed as Lightning handles the FitLoop
+        # and setting up the model, optimizer, dataloader and many more.
+
+        def fit(self):
+            optimizer = self.configure_optimizers()
+            model, optimizer = self.setup(model, optimizer)
+            dataloader = self.setup_dataloaders(self.train_dataloader())
+            model.train()
```
Comment (rohitgr7) [inline docs/source/starter/lightning_lite.rst:263]:
```suggestion
            self.model, optimizer = self.setup(self.model, optimizer)
            dataloader = self.setup_dataloaders(self.train_dataloader())
            self.model.train()
```

File: docs/source/starter/lightning_lite.rst [hunk 7]
```diff
@@ -0,0 +1,721 @@
...
+        # 3. This stays here as it belongs to the LightningModule.
+
+        def forward(self, x):
+            return self.model(x)
+
+        def training_step(self, batch, batch_idx):
+            return self.forward(batch)
+
+        def configure_optimizers(self):
+            return torch.optim.SGD(model.parameters(), ...)
```
Comment (rohitgr7) [inline docs/source/starter/lightning_lite.rst:280]:
```suggestion
            return torch.optim.SGD(self.model.parameters(), ...)
```

File: docs/source/starter/lightning_lite.rst [hunk 8]
```diff
@@ -0,0 +1,721 @@
...
+    # int: run on 2 GPUs
+    lite = Lite(devices=2, accelerator="gpu")
+
+    # list: run on GPUs 1, 4 (by bus ordering)
+    lite = Lite(devices=[1, 4], accelerator="gpu")
+    lite = Lite(devices="1, 4", accelerator="gpu")  # equivalent
+
+    # -1: run on all GPUs
+    lite = Lite(devices=-1)
+    lite = Lite(devices="-1")  # equivalent
```
Comment (rohitgr7) [inline docs/source/starter/lightning_lite.rst:427]:
```suggestion
    lite = Lite(devices=-1, accelerator="gpu")
    lite = Lite(devices="-1", accelerator="gpu")  # equivalent
```
I think accelerator is still required.

**Review by rohitgr7 — APPROVED** [review]
so cool :tada: :rocket:

File: docs/source/starter/lightning_lite.rst (hunk 2 above)
Comment (tchaton) [reply docs/source/starter/lightning_lite.rst:86]: Done !

File: docs/source/starter/lightning_lite.rst (hunk 3 above)
Comment (tchaton) [reply docs/source/starter/lightning_lite.rst:194]: Done !

File: docs/source/starter/lightning_lite.rst (hunk 4 above)
Comment (tchaton) [reply docs/source/starter/lightning_lite.rst:223]: Done !

File: docs/source/starter/lightning_lite.rst (hunk 5 above)
Comment (tchaton) [reply docs/source/starter/lightning_lite.rst:252]: Done !

File: docs/source/starter/lightning_lite.rst (hunk 6 above)
Comment (tchaton) [reply docs/source/starter/lightning_lite.rst:261]: Done !

File: docs/source/starter/lightning_lite.rst (hunk 7 above)
Comment (tchaton) [reply docs/source/starter/lightning_lite.rst:278]: Done !

File: docs/source/starter/lightning_lite.rst (hunk 8 above)
Comment (tchaton) [reply docs/source/starter/lightning_lite.rst:425]: Done !

File: docs/source/starter/lightning_lite.rst [hunk 9]
```diff
@@ -0,0 +1,719 @@
...
+:class:`~pytorch_lightning.lite.LightningLite` enables pure PyTorch users to scale their existing code
+on any kind of device while retaining full control over their own loops and optimization logic.
+
+:class:`~pytorch_lightning.lite.LightningLite` is the right tool for you if you match one of the two following descriptions:
+
+- I want to quickly scale my existing code to multiple devices with minimal code changes.
+- I would like to convert my existing code to the Lightning API, but a full path to Lightning transition might be too complex. I am looking for a stepping stone to ensure reproducibility during the transition.
+
+
+.. note:: LightningLite is a beta feature and will be considered stable for v1.6.
```
Comment (justusschock) [inline docs/source/starter/lightning_lite.rst:22]:
```suggestion
.. warning:: `LightningLite` is currently a beta feature. API is subject to change.
```

Make this a warning. I wouldn't mention when this will be stable in case we need to change. We can still target this internally.

**Review by justusschock — APPROVED** [review]

File: docs/source/starter/lightning_lite.rst (hunk 9 above)
Comment (tchaton) [reply docs/source/starter/lightning_lite.rst:22]: Done !

File: docs/source/starter/lightning_lite.rst [hunk 10]
```diff
@@ -0,0 +1,731 @@
...
+
+
+    Lite(strategy="ddp", gpus=8, num_nodes=32, accelerator="gpu").run()
+
+
+.. note:: We recommend instantiating the models within the :meth:`~pytorch_lightning.lite.LightningLite.run` method as large models would cause an out-of-memory error otherwise.
+
+.. note:
+
+    If you require a custom data or model device placement, you can deactivate
```
Comment (tchaton) [inline docs/source/starter/lightning_lite.rst:191]: @justusschock Mind checking this, it is a bit of a mouthful.

Comment (tchaton) [reply docs/source/starter/lightning_lite.rst]: Done !

File: docs/source/starter/lightning_lite.rst [hunk 11]
```diff
@@ -184,17 +184,16 @@ Here is an example while running on 256 GPUs.
     Lite(strategy="ddp", gpus=8, num_nodes=32, accelerator="gpu").run()
 
 
-.. note:: We recommend instantiating the models within the :meth:`~pytorch_lightning.lite.LightningLite.run` method as large models would cause an out-of-memory error otherwise.
+If you require a custom data or model device placement, you can deactivate
+:class:`~pytorch_lightning.lite.LightningLite` automatic placement by doing
```
Comment (justusschock) [inline docs/source/starter/lightning_lite.rst:188]:
```suggestion
:class:`~pytorch_lightning.lite.LightningLite` automatic device placement by doing
```

**Review by kaushikb11 — APPROVED** [review]
LGTM!
